test/integration/crud/abstract_operation.test.ts

```diff
@@ -0,0 +1,324 @@
+import { expect } from 'chai';
+import * as sinon from 'sinon';
+
+import { executeOperation, Long, Server, Topology } from '../../mongodb';
+import * as mongodb from '../../mongodb';
+import { topologyWithPlaceholderClient } from '../../tools/utils';
+
+describe('abstract operation', async function () {
+  describe('command name getter', async function () {
+    interface AbstractOperationSubclasses {
+      subclassCreator: () => mongodb.AbstractOperation;
+      subclassType: any;
+    }
+
+    const serverlessOperationSubclasses = [
+      'RunAdminCommandOperation',
+      'RunCommandOperation',
+      'OptionsOperation',
+      'IsCappedOperation',
+      'BulkWriteOperation',
+      'IndexExistsOperation',
+      'IndexOperation',
+      'CollectionsOperation',
+      'IndexInformationOperation'
+    ];
+
+    const sameServerOnlyOperationSubclases = ['GetMoreOperation', 'KillCursorsOperation'];
+
+    let client;
+    let db;
+    let admin;
+    let collection;
+    let constructorServer;
+    const subclassArray: AbstractOperationSubclasses[] = [
```

Comment:
First - nice work with these tests.  They're thorough and well structured.

but I was curious about the intention behind this exhaustive testing.  is it really necessary?  I generally think exhaustive testing is good when we have complex, important logic that we must ensure is correct.  This, however, is just a single property on each class.  It seems bit overkill.  

I was thinking about what would happen in the future when we add new operation classes.  A new operation class will not fail any test in this file without manual modification of this file because these tests are built from a hard-coded list of operations.

One argument in favor of exhaustive testing is that it aids future developers.  By writing an exhaustive test, if a new developer adds a new operation class that doesn't match the criteria outlined in these tests, it should fail and alert the developer.  This prevents "tribal knowledge" being important to ensuring that a codebase works as intended.  But these tests require modification when a new operation class is added and nothing would fail if a new operation class is added, _and_ a developer would need to know these tests exist.  So there's some level of "tribal knowledge" necessary regardless of whether or not we have these tests.

My 2-cents is that a descriptive comment on `get commandName()` in `AbstractOperation` is sufficient and we can remove all tests.  If we _really_ want to test this property, I'd say it's sufficient to just test string equality.  `it('AggregateOperation.commandName is `aggregate`', () => { ... })`.  Tests confirming that the string is as intended combined with a thorough comment on `get commandName()` is sufficient to make me confident that our command names are correct and won't change.

Reply:
I agree that are command names likely won't change.

The reasoning behind the exhaustive tests was to essentially ensure that the command name matches the first key of the command / to ensure I didn't make any mistakes whilst creating the getters / or miss any. I generated the subclasses by iterating through the mongodb classes and checking for subclasses of AbstractOperation, so as to ensure I didn't miss any subclasses during my implementation. So personally, I do think that the sinon test is more important than string equality, since the whole point of command name is that it matches the first key of the command.

If a new operation class is added, and we want to ensure it is in the test file, we can loop through the` subclasses of Abstract Operation === subclassesArray`, but seeing as the current tests may be overkill, that might not be favorable.

Reply:
Also, we can add a test in after logging is completed that goes like "for every commandStarted log there is a corresponding serverSelectionStarted log for a command of the same name" to ensure that commandName works for each new subclass as intended. This way we don't really need to change these tests. What do you think @baileympearson?

Reply:
> So personally, I do think that the sinon test is more important than string equality, since the whole point of command name is that it matches the first key of the command.

Are we testing this behavior now?  I see that we're checking that the command name exists in the command document, but no tests that assert it's the first key.

> If a new operation class is added, and we want to ensure it is in the test file, we can loop through the subclasses of Abstract Operation === subclassesArray, but seeing as the current tests may be overkill, that might not be favorable.

I don't fully understand what this is saying.  Are you suggesting we generate `subclassArray` programmatically?

Reply:
We're not asserting that it is the first key, we very well could. 

`subclassArray` is a bit confusingly named, it is an array of subclass names and instances. 
We can't really generate `subclassArray` in a neat programmatic way since subclasses do not have consistent constructor arguments. I'm suggesting that we can generate the names only of the subclasses programmatically, and detect if any subclasses are missing from `subclassArray`.

Reply:
I think that's probably overkill too.  I don't feel strongly here, I'll go ahead and resolve this conversation.